system-tests/test.py

```diff
@@ -13,51 +13,48 @@
 import difflib
 import logging
 from logging import StreamHandler, FileHandler
+import os.path
 from pathlib import Path
 import shutil
 import subprocess
 import sys
 import os
 import traceback
+from typing import List, Sequence
 
 import systest  # type: ignore
 from systest import Sequencer
 
 
-class FabTestCase(systest.TestCase):
-    '''Run Fab against source tree and validate result.'''
-    # The result is held in a file 'expected.txt' in the test directory.
-    #
-    # This comment exists as the framework hijacks the docstring for output.
+class RunTestCase(systest.TestCase):
+    """
+    Runs a tool from the fab collection and compares output with expected one.
+    """
 
-    def __init__(self, test_directory: Path, fpp_flags: str = None):
+    def __init__(self,
+                 test_directory: Path,
+                 working_dir: Path,
+                 expectation_file: Path,
+                 entry_point: str,
+                 args: Sequence[str] = ()):
         super().__init__(name=test_directory.stem)
         self._test_directory: Path = test_directory
-        self._fpp_flags = fpp_flags
-
-        expectation_file = test_directory / 'expected.txt'
+        self._working_dir: Path = working_dir
+        self._entry_point = entry_point
+        self._arguments = args
         self._expected = expectation_file.read_text('utf-8') \
             .splitlines(keepends=True)
 
-    def setup(self):
-        working_dir: Path = self._test_directory / 'working'
-        if working_dir.is_dir():
-            shutil.rmtree(str(working_dir))
-
-    def teardown(self):
-        working_dir: Path = self._test_directory / 'working'
-        shutil.rmtree(str(working_dir))
-
     def run(self):
-        command = ['python3', '-m', 'fab']
-        if self._fpp_flags is not None:
-            command.append('--fpp-flags=' + self._fpp_flags)
-        command.append(self._test_directory)
-
-        environment = {
-            'PYTHONPATH': 'source',
-            'PATH': os.environ.get("PATH", '')
-            }
+        script = "import sys; import fab.entry; " \
+                 f"sys.exit(fab.entry.{self._entry_point}())"
+        command = ['python3', '-c', script,
+                   '-w', self._working_dir]
+        command.extend(self._arguments)
+        new_path = os.environ.get('PATH') \
```

Comment:
This has lost the protection against the (extremely unlikely) case that `PATH` isn't set - it's probably fine but just pointing it out

Reply:
I've replaced this with a more comprehensive solution which should solve that problem. Unlikely as it is.